components/webln/nwc.js

```diff
@@ -0,0 +1,224 @@
+// https://github.com/getAlby/js-sdk/blob/master/src/webln/NostrWeblnProvider.ts
+
+import { createContext, useCallback, useContext, useEffect, useState } from 'react'
+import { Relay, finalizeEvent, nip04 } from 'nostr-tools'
+
+const NWCContext = createContext()
+
+export function NWCProvider ({ children }) {
+  const [nwcUrl, setNwcUrl] = useState('')
+  const [walletPubkey, setWalletPubkey] = useState()
+  const [relayUrl, setRelayUrl] = useState()
+  const [secret, setSecret] = useState()
+  const [enabled, setEnabled] = useState()
+  const [relay, setRelay] = useState()
+
+  const name = 'NWC'
+  const storageKey = 'webln:provider:nwc'
+
+  const loadConfig = useCallback(() => {
+    const config = window.localStorage.getItem(storageKey)
+    if (!config) return
+    const configJSON = JSON.parse(config)
+    setNwcUrl(configJSON.nwcUrl)
+  }, [])
+
+  const saveConfig = useCallback(async (config) => {
+    setNwcUrl(config.nwcUrl)
+    // XXX Even though NWC allows to configure budget,
+    // this is definitely not ideal from a security perspective.
+    window.localStorage.setItem(storageKey, JSON.stringify(config))
+  }, [])
+
+  const clearConfig = useCallback(() => {
+    window.localStorage.removeItem(storageKey)
+    setNwcUrl('')
+  }, [])
+
+  useEffect(() => {
+    let relay
+    (async function () {
+      if (relayUrl) {
+        relay = await Relay.connect(relayUrl)
+        setRelay(relay)
+      }
+    })().catch((err) => {
+      console.error(err)
+      setRelay(null)
+    })
+    return () => {
+      relay?.close()
+      setRelay(null)
+    }
+  }, [relayUrl])
```

Comment:
In c91f9f22, I switched from opening a new relay connection for every zap to having a relay connection open the whole time and reuse that one for all zaps. I did this because it's more friendly towards the resources of the relay since if we zap multiple times without confirming, multiple connections would be kept open. It's generally more efficient to reuse an existing connection, for example due to less state that needs to be maintained and less overhead of opening and closing connections.

However, this code might be less stable now since it's missing reconnect code.

I also thought about having it open only for 60 seconds and then closing it so the relay doesn't have to keep a connection open. But might be something to think about later.